File: leetcode/linked-list/445. Add Two Numbers II/py/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Time O(N)
# Space O(N)
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
      l1 = self.reverseList(l1)
      l2 = self.reverseList(l2)
      add = 0
      head = None
    
      while l1 or l2:
        x1 = l1.val if l1 else 0
        x2 = l2.val if l2 else 0
        
        total = add + x1 + x2 
        val = total % 10
        add = total // 10
        
        new_node = ListNode(val)
        new_node.next = head
        head = new_node
        
        l1 = l1.next if l1 else None
        l2 = l2.next if l2 else None
        
      if add > 0:
        new_node = ListNode(add)
        new_node.next = head
        head = new_node
        
      return head
      
        
      return dummy.next
        
    def reverseList(self, head: ListNode) -> ListNode:  
        dummy = None
        p = head
        
        while p:
            curr = ListNode(p.val)
            curr.next = dummy
            dummy = curr
            p = p.next
            
        return dummy
            
        

File: leetcode/linked-list/445. Add Two Numbers II/js/main.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

// Time O(N)
// Space O(N)

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
const addTwoNumbers = (l1, l2) => {
  l1 = reverse(l1);
  l2 = reverse(l2);
  let add = 0;
  let head = null;

  while (l1 || l2) {
    let x1 = l1 ? l1.val : 0;
    let x2 = l2 ? l2.val : 0;

    let total = x1 + x2 + add;
    let val = total % 10;
    add = Math.floor((total - val) / 10);

    let newNode = new ListNode(val);
    newNode.next = head;
    head = newNode;

    l1 = l1 ? l1.next : null;
    l2 = l2 ? l2.next : null;
  }

  if (add > 0) {
    let newNode = new ListNode(add);
    newNode.next = head;
    head = newNode;
  }

  return head;

  function reverse(node) {
    let head = null;

    while (node) {
      let newNode = new ListNode(node.val);
      newNode.next = head;

      head = newNode;

      node = node.next;
    }

    return head;
  }
};

// Time O(N)
// Space O(N)
const addTwoNumbers_II = (l1, l2) => {
  let s1 = [];
  let s2 = [];
  let headA = l1;
  let headB = l2;

  while (headA !== null) {
    s1.push(headA.val);
    headA = headA.next;
  }

  while (headB !== null) {
    s2.push(headB.val);
    headB = headB.next;
  }

  let sum = 0;
  let list = new ListNode(0);

  while (s1.length || s2.length) {
    if (s1.length) sum += s1.pop();

    if (s2.length) sum += s2.pop();

    list.val = sum % 10;
    let head = new ListNode(sum / 10);

    head.next = list;
    list = head;

    sum /= 10;
  }

  return list.val === 0 ? list.next : list;
};
